pkg/compactor/block_upload.go

```diff
@@ -262,29 +287,67 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 	w.WriteHeader(http.StatusOK)
 }
 
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) error {
+func (c *MultitenantCompactor) validateAndCompleteBlockUpload(logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta *metadata.Meta, validation func(context.Context) error) {
 	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
-	// Upload meta file so block is considered complete
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// start a go routine that updates the validation file's timestamp every heartbeat interval
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.periodicValidationUpdater(ctx, logger, blockID, userBkt, cancel)
+	}()
+
+	if err := validation(ctx); err != nil {
+		level.Error(logger).Log("msg", "error while validating block", "err", err)
+		cancel()
+		wg.Wait()
+		err := c.uploadValidationWithError(context.Background(), blockID, userBkt, err.Error())
+		if err != nil {
+			level.Error(logger).Log("msg", "error updating validation file after failed block validation", "err", err)
+		}
+		return
+	}
+
+	cancel()
+	wg.Wait() // use waitgroup to ensure validation ts update is complete
+
+	if err := c.markBlockComplete(context.Background(), logger, userBkt, blockID, meta); err != nil {
+		if err := c.uploadValidationWithError(ctx, blockID, userBkt, err.Error()); err != nil {
```

Comment:
We can't use `ctx` here. Perhaps we should rename `ctx` to `updaterCtx` to clarify its usage and avoid mistake like this.

We could also use block `{ <code> }` to reduce scope of `ctx`, `cancel` and `wg` variables, or extract first half of this function into separate function.

Reply:
Thanks for the catch, chose to scope the first half of the function then reassign `ctx`.